harness/harness-core
# feat: [DEL-4963]: Reduce Mongo footprints to update delegate heartbeat by deprecating DelegateConnection entity

## Describe your changes

Summary:
- Consolidating delegate heartbeat Mongo updates to reduce footprints

Problem Statement:
- Currently, we have 2 different ways to update heartbeat registration from the delegate, one is to update lastHeartBeat in delegate entity as well as an entry to collection DelegateConnections. The purpose behind both is to track delegate liveness
Delegate HB(HeartBeat) call frequency is one per minute per delegate.

Solution:
- Deprecate DelegateConnections usage and rely on HB from the delegate entity. This will reduce half of the HB(HeartBeat) update + read calls to mongo.  

Test:
CG
- Install CG k8,shell,docker delegate and verify HB updating in DB, verify in UI with older version of the delegate
- Install CG k8,shell,docker delegate and verify HB updating in DB, verify in UI with updated version of the delegate
- Install CG Polling mode delegate and verify HB updating in DB, verify in UI with the older version of the delegate
- Install CG Polling mode delegate and verify HB updating in DB, verify in UI with updated version of the delegate
- K8 task end-to-end flow
- Test delete delegate flow, 1)verify status in DB and UI as disconnected 2)Delete from UI when connected and disconnected
- Test HB update from 2 same location with same delegate connection id and verify self-destruct happening for both polling mode and streaming mode delegate 
- Verify HB updates reflect while delegate restart both in UI and DB
- Verify HB update on new version upgrade
- Verify listing delegate status with Delegate groups

NG
- Install CG k8, docker delegate and verify HB updating in DB, verify in UI with older version of the delegate
- Install CG k8,docker delegate and verify HB updating in DB, verify in UI with updated version of the delegate
- NG pipeline execution, task end-to-end flow
- Test delete delegate and verify status in DB and UI as disconnected
- Verify HB updates reflect while delegate restart both in UI and DB
- Verify HB update on new version upgrade
- Verify listing delegate status with Delegate groups

Immutable
-Verify upgrader, unregister and display of connected delegates in the UI

 GraphQL
- List delegate API call

## Checklist

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/42897)
